lib/src/modules/task.cc

```diff
@@ -240,30 +239,43 @@ static std::string createUrlEndpoint(const lth_jc::JsonContainer& uri) {
 //    the same task file.
 // The downloaded task file's permissions will be set to rwx for user and rx for
 // group for non-Windows OSes.
-static void downloadTaskFile(const std::vector<std::string>& master_uris,
+static bool downloadTaskFile(const std::vector<std::string>& master_uris,
                              lth_curl::client& client,
                              const fs::path& file_path,
                              const lth_jc::JsonContainer& uri) {
-    auto url = master_uris[0] + createUrlEndpoint(uri);
-    lth_curl::request req(url);
-    // timeout from connection after one minute, can configure
-    req.connection_timeout(60000);
+    auto endpoint = createUrlEndpoint(uri);
+    for (auto& master_uri : master_uris) {
+        auto url = master_uri + endpoint;
+        lth_curl::request req(url);
+        // timeout from connection after one minute, can configure
+        req.connection_timeout(60000);
+        try {
 #ifndef _WIN32
-    client.download_file(req, file_path.string(), NIX_TASK_FILE_PERMS);
+            client.download_file(req, file_path.string(), NIX_TASK_FILE_PERMS);
 #else
-    client.download_file(req, file_path.string());
+            client.download_file(req, file_path.string());
 #endif
+            break;
+        } catch (lth_curl::http_file_download_exception& e) {
+            // Do nothing here -- we want to try the next master-uri.
+
+            // TODO: Should we collect all the exception messages here to pass in to the overall
+            // Module::ProcessingError if all master_uris fail? Might not be a bad idea.
+        }
+    }
+
+    return fs::exists(file_path);
 }
 
 // This method does the following. If the file matching the "filename" field of the
 // file_obj JSON does not exist OR if its hash does not match the sha value in the
 // "sha256" field of file_obj, then:
-//    (1) The file is downloaded using Leatherman.curl from the first entry in
-//    master-uris to a temporary directory. If this download fails, a PXP error
-//    is thrown.
+//    (1) The file is downloaded using Leatherman.curl by trying each of the master_uris
+//    until one of them succeeds. If this download fails, a PXP error is thrown.
 //
 //    (2) If the downloaded file's sha does not match the provided sha, then a PXP
-//        error is returned.
+//        error is returned. TODO: Now that we are trying all the master_uris for
+//        download, should we try another master_uri if the shas do not match?
```

Comment:
IMO we should.

Reply:
What about the opposite case: Obtaining a SHA of the most recent version but downloading an older file?
In that case it would make sense to try downloading from other masters precisely for the reason you describe.
But can that happen in a normally operating infrastructure? I guess that depends on the order of the URIs in the `master-uris` configuration parameter.

Reply:
That's fair. In practice, multiple masters will generally be used for MoM/replicas, or load balancers. In the 1st case, the MoM should be up-to-date. With load balancers, you might get an out-of-date master first, then an up-to-date master. I still think it seems odd to retry just because you didn't get the answer you expected.

Reply:
I still don't think we should try unless we get feedback saying that's an issue. It's more complicated and not necessarily the right fix.